File: src/main/java/com/teamwizardry/librarianlib/api/gui/components/ComponentSprite.java
package com.teamwizardry.librarianlib.api.gui.components;

import com.teamwizardry.librarianlib.api.gui.GuiComponent;
import com.teamwizardry.librarianlib.api.gui.Option;
import com.teamwizardry.librarianlib.api.util.misc.Color;
import com.teamwizardry.librarianlib.client.Sprite;
import com.teamwizardry.librarianlib.math.Vec2;
import org.lwjgl.opengl.GL11;

public class ComponentSprite extends GuiComponent<ComponentSprite> {

	public Option<ComponentSprite, Boolean> depth = new Option<>(true);
	public Option<ComponentSprite, Color> color = new Option<>(Color.WHITE);
	
	protected Sprite sprite;
	
	public ComponentSprite(Sprite sprite, int x, int y) {
		this(sprite, x, y, sprite.getWidth(), sprite.getHeight());
	}
	
	public ComponentSprite(Sprite sprite, int x, int y, int width, int height) {
		super(x, y, width, height);
		this.sprite = sprite;
	}
	
	@Override
	public void drawComponent(Vec2 mousePos, float partialTicks) {
		boolean alwaysTop = !depth.getValue(this);
		
		if(alwaysTop) {
			// store the current depth function
			GL11.glPushAttrib(GL11.GL_DEPTH_BUFFER_BIT);

			// by using GL_ALWAYS instead of disabling depth it writes to the depth buffer
			// imagine a mountain, that is the depth buffer. this causes the sprite to write
			// it's value to the depth buffer, cutting a hole down wherever it's drawn.
			GL11.glDepthFunc(GL11.GL_ALWAYS);
		}
		color.getValue(this).glColor();
		sprite.getTex().bind();
		sprite.draw(pos.xf, pos.yf, size.xi, size.yi);
		if(alwaysTop)
			GL11.glPopAttrib();
	}

	public Sprite getSprite() {
		return sprite;
	}

	public void setSprite(Sprite sprite) {
		this.sprite = sprite;
	}
	
}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/components/ComponentSprite.kt
package com.teamwizardry.librarianlib.client.gui.components

import com.teamwizardry.librarianlib.client.gui.GuiComponent
import com.teamwizardry.librarianlib.client.gui.Option
import com.teamwizardry.librarianlib.client.sprite.Sprite
import com.teamwizardry.librarianlib.client.util.glColor
import com.teamwizardry.librarianlib.common.util.event.Event
import com.teamwizardry.librarianlib.common.util.math.Vec2d
import org.lwjgl.opengl.GL11
import java.awt.Color

class ComponentSprite @JvmOverloads constructor(var sprite: Sprite?, x: Int, y: Int, width: Int = sprite?.width ?: 16, height: Int = sprite?.height ?: 16) : GuiComponent<ComponentSprite>(x, y, width, height) {

    class AnimationLoopEvent(val component: ComponentSprite) : Event()


    var depth = Option<ComponentSprite, Boolean>(true)
    var color = Option<ComponentSprite, Color>(Color.WHITE)

    var lastAnim: Int = 0

    override fun drawComponent(mousePos: Vec2d, partialTicks: Float) {
        val alwaysTop = !depth.getValue(this)
        val sp = sprite
        sp ?: return
        if (alwaysTop) {
            // store the current depth function
            GL11.glPushAttrib(GL11.GL_DEPTH_BUFFER_BIT)

            // by using GL_ALWAYS instead of disabling depth it writes to the depth buffer
            // imagine a mountain, that is the depth buffer. this causes the sprite to write
            // it's value to the depth buffer, cutting a hole down wherever it's drawn.
            GL11.glDepthFunc(GL11.GL_ALWAYS)
        }
        if(sp.frameCount > 0 && lastAnim / sp.frameCount < animationTicks / sp.frameCount) {
            BUS.fire(AnimationLoopEvent(this))
        }
        lastAnim = animationTicks
        color.getValue(this).glColor()
        sp.tex.bind()
        sp.draw(animationTicks, pos.xf, pos.yf, size.xi.toFloat(), size.yi.toFloat())
        if (alwaysTop)
            GL11.glPopAttrib()
    }

}
